File: app/src/main/java/ua/hospes/rtm/data/race/RaceRepositoryImpl.java
package ua.hospes.rtm.data.race;

import android.content.ContentValues;
import android.util.Pair;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.functions.BiFunction;
import ua.hospes.rtm.data.race.mapper.RaceMapper;
import ua.hospes.rtm.data.race.operations.UpdateRaceOperation;
import ua.hospes.rtm.data.race.storage.RaceDbStorage;
import ua.hospes.rtm.data.sessions.SessionsRepositoryImpl;
import ua.hospes.rtm.data.sessions.models.SessionDb;
import ua.hospes.rtm.data.sessions.storage.SessionsDbStorage;
import ua.hospes.rtm.domain.race.RaceRepository;
import ua.hospes.rtm.domain.race.models.RaceItem;
import ua.hospes.rtm.domain.race.models.RaceItemDetails;
import ua.hospes.rtm.domain.sessions.Session;
import ua.hospes.rtm.domain.sessions.SessionsRepository;
import ua.hospes.rtm.domain.team.Team;
import ua.hospes.rtm.domain.team.TeamsRepository;
import ua.hospes.rtm.utils.Optional;

/**
 * @author Andrew Khloponin
 */
@Singleton
public class RaceRepositoryImpl implements RaceRepository {
    private final RaceDbStorage raceDbStorage;
    private final TeamsRepository teamsRepository;
    private final SessionsDbStorage sessionsDbStorage;
    private final SessionsRepository sessionsRepository;


    @Inject
    RaceRepositoryImpl(RaceDbStorage raceDbStorage, TeamsRepository teamsRepository, SessionsDbStorage sessionsDbStorage, SessionsRepositoryImpl sessionsRepository) {
        this.raceDbStorage = raceDbStorage;
        this.teamsRepository = teamsRepository;
        this.sessionsDbStorage = sessionsDbStorage;
        this.sessionsRepository = sessionsRepository;
    }


    @Override
    public Observable<RaceItem> get() {
        return raceDbStorage.get()
                .flatMapSingle(raceItemDb -> Single.zip(Single.just(raceItemDb), getTeamById(raceItemDb.getTeamId()), getSessionById(raceItemDb.getSessionId()), RaceMapper::map))
                .flatMapSingle(raceItem -> Single.zip(Single.just(raceItem), getRawSessionsByTeamId(raceItem.getTeam().getId()), new CalculateRaceItemDetails()));
    }

    @Override
    public Observable<List<RaceItem>> listen() {
        return raceDbStorage.listen()
                .flatMap(raceItemDbs -> Observable.fromIterable(raceItemDbs)
                        .flatMapSingle(raceItemDb -> Single.zip(Single.just(raceItemDb), getTeamById(raceItemDb.getTeamId()), getSessionById(raceItemDb.getSessionId()), RaceMapper::map))
                        .flatMapSingle(raceItem -> Single.zip(Single.just(raceItem), getRawSessionsByTeamId(raceItem.getTeam().getId()), new CalculateRaceItemDetails()))
                        .toList().toObservable()
                );
    }

    @Override
    public Observable<RaceItem> listen(int id) {
        return raceDbStorage.listen(id)
                .flatMap(raceItemDbs -> Observable.fromIterable(raceItemDbs)
                        .flatMapSingle(raceItemDb -> Single.zip(Single.just(raceItemDb), getTeamById(raceItemDb.getTeamId()), getSessionById(raceItemDb.getSessionId()), RaceMapper::map))
                        .flatMapSingle(raceItem -> Single.zip(Single.just(raceItem), getRawSessionsByTeamId(raceItem.getTeam().getId()), new CalculateRaceItemDetails()))
                        .toList().toObservable()
                )
                .flatMapIterable(raceItems -> raceItems);
    }

    @Override
    public Observable<Boolean> addNew(RaceItem... items) {
        return Observable.fromArray(items)
                .map(RaceMapper::map)
                .toList()
                .flatMapObservable(raceDbStorage::add)
                .map(result -> result.getResult() > 0);
    }

    @Override
    public Observable<Boolean> update(List<RaceItem> items) {
        return Observable.fromIterable(items)
                .map(RaceMapper::map)
                .map(UpdateRaceOperation::new)
                .toList()
                .flatMapObservable(raceDbStorage::updateRaces);
    }

    @Override
    public Observable<Boolean> updateByTeamId(Iterable<Pair<Integer, ContentValues>> items) {
        return Observable.fromIterable(items)
                .map(UpdateRaceOperation::new)
                .toList()
                .flatMapObservable(raceDbStorage::updateRaces);
    }

    @Override
    public Single<Integer> remove(RaceItem item) {
        return raceDbStorage.remove(RaceMapper.map(item));
    }


    @Override
    public Observable<Void> reset() {
        return raceDbStorage.reset();
    }

    @Override
    public Single<Integer> removeAll() {
        return raceDbStorage.removeAll();
    }


    private Single<Optional<Team>> getTeamById(int id) {
        return teamsRepository.get(id).map(Optional::of);
    }

    private Single<Optional<Session>> getSessionById(int id) {
        return sessionsRepository.get(id).map(Optional::of).single(Optional.empty());
    }

    private Single<List<SessionDb>> getRawSessionsByTeamId(int teamId) {
        return sessionsDbStorage.getByTeamId(teamId).toList();
    }

    private class CalculateRaceItemDetails implements BiFunction<RaceItem, List<SessionDb>, RaceItem> {
        @Override
        public RaceItem apply(RaceItem item, List<SessionDb> sessions) throws Exception {
            RaceItemDetails details = new RaceItemDetails();
            int pitStops = -1;
            for (SessionDb session : sessions) {
                pitStops += Session.Type.TRACK.name().equals(session.getType()) ? 1 : 0;
                if (session.getStartDurationTime() == -1 || session.getEndDurationTime() == -1) continue;
                if (session.getDriverId() == -1) continue;
                long duration = session.getEndDurationTime() - session.getStartDurationTime();
                details.addDriverDuration(session.getDriverId(), duration);
            }
            details.setPitStops(pitStops);
            item.setDetails(details);
            return item;
        }
    }
}

File: app/src/main/java/ua/hospes/rtm/data/race/RaceRepositoryImpl.kt
package ua.hospes.rtm.data.race

import android.content.ContentValues
import android.util.Pair

import javax.inject.Inject
import javax.inject.Singleton

import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.functions.BiFunction
import ua.hospes.rtm.data.race.mapper.RaceMapper
import ua.hospes.rtm.data.race.operations.UpdateRaceOperation
import ua.hospes.rtm.data.race.storage.RaceDbStorage
import ua.hospes.rtm.data.sessions.SessionsRepositoryImpl
import ua.hospes.rtm.data.sessions.models.SessionDb
import ua.hospes.rtm.data.sessions.storage.SessionsDbStorage
import ua.hospes.rtm.domain.race.RaceRepository
import ua.hospes.rtm.domain.race.models.RaceItem
import ua.hospes.rtm.domain.race.models.RaceItemDetails
import ua.hospes.rtm.domain.sessions.Session
import ua.hospes.rtm.domain.sessions.SessionsRepository
import ua.hospes.rtm.domain.team.Team
import ua.hospes.rtm.domain.team.TeamsRepository
import ua.hospes.rtm.utils.Optional

/**
 * @author Andrew Khloponin
 */
@Singleton
class RaceRepositoryImpl @Inject
internal constructor(private val raceDbStorage: RaceDbStorage, private val teamsRepository: TeamsRepository, private val sessionsDbStorage: SessionsDbStorage, sessionsRepository: SessionsRepositoryImpl) : RaceRepository {
    private val sessionsRepository: SessionsRepository


    init {
        this.sessionsRepository = sessionsRepository
    }


    override fun get(): Observable<RaceItem> {
        return raceDbStorage.get()
                .flatMapSingle { raceItemDb -> Single.zip<RaceItemDb, Optional<Team>, Optional<Session>, RaceItem>(Single.just<RaceItemDb>(raceItemDb), getTeamById(raceItemDb.teamId), getSessionById(raceItemDb.sessionId), Function3<RaceItemDb, Optional<Team>, Optional<Session>, RaceItem> { db, team, session -> RaceMapper.map(db, team, session) }) }
                .flatMapSingle { raceItem -> Single.zip(Single.just(raceItem), getRawSessionsByTeamId(raceItem.team.id!!), CalculateRaceItemDetails()) }
    }

    override fun listen(): Observable<List<RaceItem>> {
        return raceDbStorage.listen()
                .flatMap { raceItemDbs ->
                    Observable.fromIterable<RaceItemDb>(raceItemDbs)
                            .flatMapSingle { raceItemDb -> Single.zip<RaceItemDb, Optional<Team>, Optional<Session>, RaceItem>(Single.just<RaceItemDb>(raceItemDb), getTeamById(raceItemDb.teamId), getSessionById(raceItemDb.sessionId), Function3<RaceItemDb, Optional<Team>, Optional<Session>, RaceItem> { db, team, session -> RaceMapper.map(db, team, session) }) }
                            .flatMapSingle { raceItem -> Single.zip(Single.just(raceItem), getRawSessionsByTeamId(raceItem.team.id!!), CalculateRaceItemDetails()) }
                            .toList().toObservable()
                }
    }

    override fun listen(id: Int): Observable<RaceItem> {
        return raceDbStorage.listen(id)
                .flatMap { raceItemDbs ->
                    Observable.fromIterable<RaceItemDb>(raceItemDbs)
                            .flatMapSingle { raceItemDb -> Single.zip<RaceItemDb, Optional<Team>, Optional<Session>, RaceItem>(Single.just<RaceItemDb>(raceItemDb), getTeamById(raceItemDb.teamId), getSessionById(raceItemDb.sessionId), Function3<RaceItemDb, Optional<Team>, Optional<Session>, RaceItem> { db, team, session -> RaceMapper.map(db, team, session) }) }
                            .flatMapSingle { raceItem -> Single.zip(Single.just(raceItem), getRawSessionsByTeamId(raceItem.team.id!!), CalculateRaceItemDetails()) }
                            .toList().toObservable()
                }
                .flatMapIterable { raceItems -> raceItems }
    }

    override fun addNew(vararg items: RaceItem): Observable<Boolean> {
        return Observable.fromArray(*items)
                .map<RaceItemDb>(Function<RaceItem, RaceItemDb> { RaceMapper.map(it) })
                .toList()
                .flatMapObservable<InsertResult<RaceItemDb>>(Function<List<RaceItemDb>, ObservableSource<out InsertResult<RaceItemDb>>> { raceDbStorage.add(it) })
                .map { result -> result.getResult() > 0 }
    }

    override fun update(items: List<RaceItem>): Observable<Boolean> {
        return Observable.fromIterable(items)
                .map<RaceItemDb>(Function<RaceItem, RaceItemDb> { RaceMapper.map(it) })
                .map<UpdateRaceOperation>(Function<RaceItemDb, UpdateRaceOperation> { UpdateRaceOperation(it) })
                .toList()
                .flatMapObservable(Function<List<UpdateRaceOperation>, ObservableSource<out Boolean>> { raceDbStorage.updateRaces(it) })
    }

    override fun updateByTeamId(items: Iterable<Pair<Int, ContentValues>>): Observable<Boolean> {
        return Observable.fromIterable(items)
                .map<UpdateRaceOperation>(Function<Pair<Int, ContentValues>, UpdateRaceOperation> { UpdateRaceOperation(it) })
                .toList()
                .flatMapObservable(Function<List<UpdateRaceOperation>, ObservableSource<out Boolean>> { raceDbStorage.updateRaces(it) })
    }

    override fun remove(item: RaceItem): Single<Int> {
        return raceDbStorage.remove(RaceMapper.map(item))
    }


    override fun reset(): Observable<Void> {
        return raceDbStorage.reset()
    }

    override fun removeAll(): Single<Int> {
        return raceDbStorage.removeAll()
    }


    private fun getTeamById(id: Int): Single<Optional<Team>> {
        return teamsRepository.get(id).map(Function<Team, Optional<Team>> { Optional.of(it) })
    }

    private fun getSessionById(id: Int): Single<Optional<Session>> {
        return sessionsRepository.get(id).map<Optional<Session>>(Function<Session, Optional<Session>> { Optional.of(it) }).single(Optional.empty())
    }

    private fun getRawSessionsByTeamId(teamId: Int): Single<List<SessionDb>> {
        return sessionsDbStorage.getByTeamId(teamId).toList()
    }

    private inner class CalculateRaceItemDetails : BiFunction<RaceItem, List<SessionDb>, RaceItem> {
        @Throws(Exception::class)
        override fun apply(item: RaceItem, sessions: List<SessionDb>): RaceItem {
            val details = RaceItemDetails()
            var pitStops = -1
            for (session in sessions) {
                pitStops += if (Session.Type.TRACK.name == session.type) 1 else 0
                if (session.startDurationTime == -1 || session.endDurationTime == -1) continue
                if (session.driverId == -1) continue
                val duration = session.endDurationTime - session.startDurationTime
                details.addDriverDuration(session.driverId, duration)
            }
            details.pitStops = pitStops
            item.details = details
            return item
        }
    }
}